docs/page-objects/layouts.rst

```diff
@@ -0,0 +1,170 @@
+an unmaintainable bowl of spaghetti code?
+
+
+Mapping webpage layouts
+=======================
+
+Once you have written a :ref:`page object class <page-objects>` for a webpage
+layout, you need to make it so that your page object class is used for webpages
+that use that webpage layout.
+
+URL patterns
+------------
+
+Webpage layouts are often associated to specific URL patterns. For example, all
+the product detail pages of an e-commerce website usually have similar URLs,
+such as ``https://example.com/product/<product ID>``.
+
+When that is the case, you can :ref:`associate your page object class to the
+corresponding URL pattern <rules-intro>`.
+
+
+.. _multi-layout:
+
+Multi-layout page object classes
+--------------------------------
+
+Sometimes it is impossible to know, based on the target URL, which webpage
+layout you are getting. For example, during `A/B testing`_, you could get a
+random webpage layout on every request.
+
+.. _A/B testing: https://en.wikipedia.org/wiki/A/B_testing
+
+For these scenarios, we recommend that you create different page object classes
+for the different layouts that you may get, and then write a special
+“multi-layout” page object class, and use it to select the right page object
+class at run time based on the input you receive.
+
+Your multi-layout page object class should:
+
+#.  Declare attributes for the input that you will need to determine which page
+    object class to use.
+
+    For example, declare an :class:`HttpResponse` attribute to select a page
+    object class based on the response content.
+
+#.  Declare an attribute for every page object class that you may use depending
+    on which webpage layout you get from the target website.
+
+    They all should return the same type of :ref:`item <item-classes>` as your
+    multi-layout page object class.
+
+    Note that all inputs of all those page object classes will be resolved and
+    requested along with the input of your multi-layout page object class. For
+    example, if one page object class requires browser HTML as input, while
+    another requires an HTTP response, your multi-layout page object class asks
+    for both inputs.
+
+    If combining different inputs is a problem, consider refactoring your page
+    object classes to require similar inputs.
+
+#.  On its :meth:`~web_poet.pages.ItemPage.to_item` method:
+
+    #.  Determine, based on inputs, which page object to use.
+
+    #.  Return the output of the :meth:`~web_poet.pages.ItemPage.to_item`
+        method of that page object.
+
+You may use :class:`~web_poet.pages.MultiLayoutPage` as a base class for your
+multi-layout page object class, so you only need to implement the
+:class:`~web_poet.pages.MultiLayoutPage.layout` method that determines which
+page object to use. For example:
+
+.. code-block:: python
+
+    import attrs
+    from web_poet import handle_urls, HttpResponse, ItemPage, MultiLayoutPage, WebPage
+
+
+    @attrs.define
```

Comment:
```suggestion
```

Reply:
I take it from your feedback that `@attrs.define` is not needed when inheriting from a class that already uses it. But in this specific case, with no parent class, isn’t it needed?

Reply:
Hmm, it seems `@attrs.define` is also needed when new attributes are defined, otherwise you get things like:

```
TypeError: object.__new__() takes exactly one argument (the type to instantiate)
TypeError: __init__() got an unexpected keyword argument 'price'
```

Reply:
Hmm, it seems `@attrs.define` is also needed when new attributes are defined, otherwise you get things like:

```
TypeError: object.__new__() takes exactly one argument (the type to instantiate)
TypeError: __init__() got an unexpected keyword argument 'price'
```

Reply:
Ahhh yes, I marked this incorrectly while in a `@attrs.define` removal spree yesterday. 😅